lib/network-layer/src/Cardano/Wallet/Network/Streaming.hs

```diff
@@ -0,0 +1,165 @@
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Cardano.Wallet.Network.Streaming
+    ( withStreamingFromBlockChain
+    , TxWithContext
+    , OrRollback (..)
+    , FromBlockChain
+    , eraBlockS
+    , eraTxS
+    )
+where
+
+import Prelude
+
+import Cardano.Wallet.Network
+    ( ChainFollowLog
+    , ChainFollower (..)
+    , NetworkLayer (..)
+    )
+import Cardano.Wallet.Read
+    ( BHeader
+    , Block
+    , ChainPoint (..)
+    , ChainTip
+    , ConsensusBlock
+    , EraValue
+    , IsEra (..)
+    , K (..)
+    , Tx
+    , applyEraFunValue
+    , fromConsensusBlock
+    , getEraTransactions
+    , sequenceEraValue
+    , (:*:) (..)
+    , (:.:) (..)
+    )
+import Cardano.Wallet.Read.Block.BHeader
+    ( getEraBHeader
+    )
+import Control.Monad
+    ( forM_
+    )
+import Control.Monad.Trans.Cont
+    ( ContT (..)
+    )
+import Control.Tracer
+    ( Tracer
+    )
+import Data.List.NonEmpty
+    ( NonEmpty
+    )
+import Ouroboros.Consensus.Cardano.Block
+    ( CardanoBlock
+    , StandardCrypto
+    )
+import Streaming
+    ( MonadIO (..)
+    , Of
+    , Stream
+    )
+import UnliftIO
+    ( atomically
+    , newEmptyTMVarIO
+    , putTMVar
+    , takeTMVar
+    )
+import UnliftIO.Async
+    ( withAsync
+    )
+
+import qualified Streaming.Prelude as S
+
+-- | An era-value of a transaction with its context. The context is a pair of
+-- the chain tip and the block header.
+type TxWithContext = EraValue ((K ChainTip :*: BHeader) :*: Tx)
+
+-- | A value that can be either an element or a rollback placeholder
+data OrRollback a = Element a | Rollback ChainPoint
+    deriving (Functor, Show, Eq)
+
+-- | A stream of 'OrRollback a' values.
+-- This is a bit unfortunate, as the real interface for rollback should be
+-- 'type FromBlockChain a = Stream (ORollback a) IO'
+-- on a functor like
+-- 'data ORollback a x = Element a x | Rollback ChainPoint (ChainPoint -> x)'
+-- but this is not possible with the current NetworkLayer
+-- Instead we simply signal that a rolled back happened without actually
+-- controlling the rollback.
```

Comment:
Controlling the position after a the rollback is precisely what we cannot do with the `streaming` interface. 😛 — We have to accept whatever `ChainPoint` the rollback goes to.

This is not a bad idea — going back only as far as needed is efficient.

However, implementations that are based on checkpoints may want to roll back further to the next checkpoint; such implementations can't work with a streaming interface. But that's ok, they just have to work with a `ChainFollower` directly, and this means that the streaming interface is an additional interface, but not a substitute for the `ChainFollower`.